docs/lazy-adr/adr-005-da-modules-and-components.md

```diff
@@ -0,0 +1,122 @@
+`NamespacedShare`, leaving the process of preparing shares of any data to user.
+
+### Interface
+```go
+// DataAvailability defines an interface for Data Availability related operations over the network.
+// The goal of this interface is to become general enough to later abstract over other DA schemas, e.g. Kate comms,
+// along with different networking schemas.
+// NOTE: Put and Get relies on DAHeader which is used to address and verify integrity of data, and its solely a user's
+// responsibility to manage header persistence and transportation between peers interested in data.
+type DataAvailability interface {
+	// Put serializes Data, computes DAHeader and makes it available over the network.
+	Put(context.Context, *types.Block) (*types.DataAvailabilityHeader, error)
+
+	// Get efficiently retrieves block's Data using given DAHeader from the network.
+	Get(context.Context, *types.DataAvailabilityHeader) (*types.Data, error)
+
+	// Validate performs data availability validation over the network.
+	// If succeeds, the data is proven to be available for specified validation options.
+	Validate(context.Context, *types.DataAvailabilityHeader, ...ValidationOption) error
+}
+```
+
+### Pull DA Components
+Pull DA is the first implementation of DA interface. We already implemented it partially. The main idea behind it,
+is in its name - pull. That is, pull DA pulls Data by requesting it from the "network" mainly using
+[kDHT from IPFS/libp2p](https://github.com/libp2p/go-libp2p-kad-dht).
+
+LL's specific DHT usage is described in high-level [here](https://github.com/lazyledger/lazyledger-core/issues/395).
+
+All pull related types and logic is going to be implemented in it's own package `dah/pull`
+
+#### DAProvider
+For DHT related operation a `DAProvider` is introduced with further semantics:
+```go
+type DAProvider interface {
+    // Provide announces to the network that the node calling it hosts data addressed by given DAHeader.
+    // Announcement lives for 24h by default.
+    Provide(context.Context, DAHeader) error
+    // Reprovide periodically(every 24h by default) executes Provide over given DAHeader.
+    // It persists reporvide requirement in datastore to continue reproviding after app restarts.
+    Reprovide(DAHeader) error
+    // Unprovide prevents further reprovides for a given DAHeader
+    Unprovide(DAHeader) error
+}
+```
+
+// TODO: Add to specs as well
+We also define several system-wide parameters related to Provider and DHT:
+* `provideValidity=24h` - lifespan of a providing record.
+* `kBucketSize=20` - configures bucket size of the kademlia routing table.
+* `resiliency=3` - the number of peers closest to a target that must have responded in order for a given DHT query to
+  complete.
+* `alpha=32` - configures the number of concurrent requests for a given DHT query.
+* `protocolPrefix="/celestia"` - protocol prefix to distinguish our DHT network
```

Comment:
Is there a rationale for parameter choices?

Reply:
From what I can tell, these are mostly the current ipfs defaults (besides the protocolPrefix and I'm not sure about the resilience param). I'm sure Hlib just wrote them down to have something explicitly mentioned here to work with.

Reply:
I also increased `alpha`